### PR TITLE
Add backup system for minecolonies.dat [#526]

### LINKED ISSUE #526: Backup the minecolonies.dat [$10 awarded]
It would be nice if the minecolonnies.dat was backup automatically at each start of the server/world.

- Backup the dat file as minecolonnies-YYYY-MM-DD.dat and continue to use minecolonnies.dat 
- Do it when the server start
- add a command to do the back so server admin could schedule an automatic backup even when the server is not restarted

[Bountysource](https://www.bountysource.com/issues/41473573-backup-the-minecolonnies-dat?utm_source=41473573&utm_medium=shield&utm_campaign=ISSUE_BADGE)

<bountysource-plugin>

---
The **[$10 bounty](https://www.bountysource.com/issues/41473573-backup-the-minecolonies-dat?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** on this issue has been claimed at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by asiekierka**
Closes #526

# Changes proposed in this pull request:
- Adds backup system for minecolonies.dat as proposed in issue #526

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/commands/BackupCommand.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    @NotNull
+    @Override
+    public List<String> getTabCompletionOptions(
+                                                 @NotNull final MinecraftServer server,
+                                                 @NotNull final ICommandSender sender,
+                                                 @NotNull final String[] args,
+                                                 @Nullable final BlockPos pos)
+    {
+        return null;
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/commands/BackupCommand.java]: this is marked notnull, return an empty list instead

File: src/main/java/com/minecolonies/coremod/commands/BackupCommand.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    @Override
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
+    {
+        if (!sender.canCommandSenderUseCommand(4, "minecolonies backup"))
+        {
+            sender.addChatMessage(new TextComponentString(NO_PERMISSION_MESSAGE));
+        }
+        else
+        {
+            if (ColonyManager.backupColonyData())
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/commands/BackupCommand.java]: this should be done in the Server Thread I guess

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 3]
```diff
@@ -634,6 +684,8 @@ public static void onWorldLoad(@NotNull final World world)
         {
             if (numWorldsLoaded == 0)
             {
+                backupColonyData();
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: the result is ignored. We should print out something, depending on if it worked or not

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
some minor things

File: src/main/java/com/minecolonies/coremod/commands/BackupCommand.java (hunk 2 above)
Comment (asiekierka) [reply src/main/java/com/minecolonies/coremod/commands/BackupCommand.java]: Ah, yes, I forgot about RCON. sorry!

**Comment by marvin-bitterlich** [comment]
:shipit: @Raycoms 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/613/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/coremod/commands/BackupCommand.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    public BackupCommand(@NotNull final String... parents)
+    {
+        super(parents);
+    }
+
+    @Override
+    public void execute(@NotNull final MinecraftServer server, @NotNull final ICommandSender sender, @NotNull final String... args) throws CommandException
+    {
+        if (!sender.canCommandSenderUseCommand(4, "minecolonies backup"))
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/commands/BackupCommand.java:39]: @asiekierka We need this to be 3 it seems. There should be a constant somewhere in the commands hirarchy for that. I guess even a config option.

mc is weird xD

**Pull request merged**